File: src/embedded-cdc/dev/cdc-service/src/main/java/dev/embeddedcdc/domain/model/Member.java
package dev.embeddedcdc.domain.model;

import java.time.OffsetDateTime;

/**
 * target 의 member 한 행. grade 를 참조하는 자식 테이블이다.
 *
 * <b>gradeId 는 그냥 long 이다 — Grade 참조가 아니다.</b>
 * 이벤트는 테이블마다 따로 오므로 member 이벤트를 받는 시점에 해당 grade 가
 * target 에 있다는 보장이 없다. 객체 그래프를 만들려 들면 그 순간 조회가 필요해지고,
 * 없으면 실패한다. CDC 는 행 단위 복제라 관계는 값(외래 키 컬럼)으로만 옮긴다.
 */
public record Member(
        long id,
        String email,
        String name,
        long gradeId,
        int point,
        OffsetDateTime createdAt,
        OffsetDateTime updatedAt,
        long sourceLsn
) {
}

File: src/embedded-cdc-go/dev/cdc-service/internal/infrastructure/persistence/member.go
package persistence

import (
	"context"

	"github.com/embedded-cdc-go/cdc-service/internal/domain/model"
	"github.com/embedded-cdc-go/cdc-service/internal/domain/port"
)

// MemberRepository 는 grade 존재 여부를 확인하지 않는다 —
// 확인해도 할 수 있는 일이 실패시키기뿐이고, target 에 FK 가 없어 DB 도 막지 않는다.
// 어긋남은 대사(V6)가 잡는다.
//
// 충돌 판정 컬럼이 id 뿐이라는 점에 유의할 것. source 의 email 에는 UNIQUE 가 있지만
// target 에는 걸지 않았다 — 소프트 삭제로 행이 남아 있는 상태에서 같은 이메일이
// 새 id 로 다시 들어오면 UNIQUE 위반으로 적재가 막히기 때문이다.
type MemberRepository struct {
	store *Store
}

func NewMemberRepository(store *Store) *MemberRepository {
	return &MemberRepository{store: store}
}

var _ port.MemberRepository = (*MemberRepository)(nil)

const memberUpsertSQL = `
INSERT INTO member (id, email, name, grade_id, point, created_at, updated_at,
                    deleted, source_lsn, synced_at)
VALUES ($1, $2, $3, $4, $5, $6, $7, false, $8, now())
ON CONFLICT (id) DO UPDATE SET
    email      = EXCLUDED.email,
    name       = EXCLUDED.name,
    grade_id   = EXCLUDED.grade_id,
    point      = EXCLUDED.point,
    created_at = EXCLUDED.created_at,
    updated_at = EXCLUDED.updated_at,
    deleted    = EXCLUDED.deleted,
    source_lsn = EXCLUDED.source_lsn,
    synced_at  = EXCLUDED.synced_at
WHERE EXCLUDED.source_lsn > member.source_lsn`

func (r *MemberRepository) UpsertIfNewer(ctx context.Context, member model.Member) (int64, error) {
	lsn, err := lsnToInt64(member.SourceLSN)
	if err != nil {
		return 0, err
	}
	tag, err := r.store.exec(ctx).Exec(ctx, memberUpsertSQL,
		member.ID, member.Email, member.Name, member.GradeID, member.Point,
		member.CreatedAt, member.UpdatedAt, lsn)
	if err != nil {
		return 0, err
	}
	return tag.RowsAffected(), nil
}

func (r *MemberRepository) SoftDelete(ctx context.Context, id int64, lsn uint64) (int64, error) {
	return softDelete(ctx, r.store, "member", id, lsn)
}
